src/data/real/basic.lean

```diff
@@ -481,6 +481,29 @@ begin
   { exact csupr_const },
 end
 
+lemma exists_of_le_supr_subtype {α : Type*} {p : α → Prop}
+  (hp : {x | p x}.finite) (hp' : {x | p x}.nonempty) {f : α → ℝ} {ε : ℝ}
+  (h : ε ≤ ⨆ x : {x // p x}, f x) :
+  ∃ x, p x ∧ ε ≤ f x :=
+begin
+  haveI : nonempty {x : α // p x} := {x : α | p x}.nonempty_coe_sort.2 hp',
+  haveI : fintype {x // p x} := hp.fintype,
+  have : (⨆ x : {x // p x}, f x) = hp.to_finset.sup' (hp.nonempty_to_finset.2 hp') f,
```

Comment:
This seems to be a general statement, useful to go between finsets and finite sets. Could you write it as a general lemma? But in general I have the impression that you are a little bit fighting against the system, and that many things would be smoother if you just used finsets everywhere (and wrote your supremums not using subtypes, which are always painful, but simply `⨆ x ∈ s, f x` (or even `s.sup' ... f`).

Reply:
I think `⨆ x ∈ s, f x` is a bad spelling for `csupr` because `(⨆ x ∈ ({-1} : set ℝ), id x) = 0` while I would expect it to be `-1`. The spelling with subtypes do not have this problem. I will change the spelling to use `finset.sup'` however

Reply:
Ah yes, you're right that the spelling `⨆ x ∈ s, f x` is bad! Between the other two (subtypes and `finset.sup'`), the question is really which one is easier to use. I suspect it should be `finset.sup'`, but we may have surprises here. Since you have experimented with both, can you tell which you think is nicer in the end? (And that's the one we should use).

Reply:
I found `finset.sup'` to be easier to use since there is not that much APIs for `csupr`. But I think I'm lucky in the case since `finset.range` is readily available. I would imagine it being much harder if I'm taking the supremum over a more complicated set for which I'll have to construct a finset and prove its bounded above each time (or define a new definition for said set)